### PR TITLE
Streamline data clean up

### PR DESCRIPTION
**Pull request by mkangia**
[Ticket](https://dimagi-dev.atlassian.net/browse/ICDS-952)
[Spec](https://docs.google.com/document/d/15UBorQURkXVNv4kYj-cpaU6230Dh0Ggg0DXm_G4t3mE/edit#)
Next steps
1) have logging to note down case ids getting updated
2) After few manual runs when things look stable and reliable, make it automatic to run periodically every 3 months

##### FEATURE FLAG
`RUN_DATA_MANAGEMENT_TASKS`

Best review by commit
Buddy: @millerdev

### CODE REVIEW

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 1]
```diff
@@ -761,6 +762,8 @@ def extra_filters(self, for_count=False):
             filters.append(Q(server_modified_on__lt=self.end_date))
         if self.case_type is not None:
             filters.append(Q(type=self.case_type))
+        if self.case_ids:
+            filters.append(Q(case_id__in=self.case_ids))
```
Comment (snopoke) [inline corehq/form_processor/backends/sql/dbaccessors.py]: Rather than add this here you should use `CaseAccessor.iter_cases`

File: custom/icds/data_management/base.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        processor = BulkDocProcessor(record_provider, self.doc_processor(self.domain))
+        processor.run()
+
+
+class SQLBasedDataManagement(DataManagement):
+    def __init__(self, domain, db_alias, from_date=None, till_date=None):
+        super().__init__(domain)
+        self.db_aliases = [db_alias] if db_alias else None
+        self.from_date = from_date
+        self.till_date = till_date
```
Comment (snopoke) [inline custom/icds/data_management/base.py]: nit: common convention for date naming is `start_date`, `end_date`

File: custom/icds/data_management/models.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                self.status = self.STATUS_SUCCESS
+            finally:
+                self.ended_on = datetime.utcnow()
+                self.save()
+
+    def _set_in_progress(self):
+        self.status = self.STATUS_IN_PROGRESS
+        self.started_on = datetime.utcnow()
+        self.save()
+        self.refresh_from_db()
```
Comment (snopoke) [inline custom/icds/data_management/models.py]: why is this necessary?

File: custom/icds/data_management/models.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    error = models.TextField(blank=True, null=True)
+    status = models.SmallIntegerField(choices=STATUS_CHOICES, default=STATUS_PENDING)
+
+    def execute(self):
+        """
+        run the updates through DataManagement
+        """
+        self._set_in_progress()
+        if self.slug not in DATA_MANAGEMENT_TASKS:
+            self._note_error("Unexpected slug %s" % self.slug)
```
Comment (snopoke) [inline custom/icds/data_management/models.py]: surely this should be handled earlier

File: custom/icds/data_management/views.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+@location_safe
+@method_decorator([login_and_domain_required,
+                   toggles.RUN_DATA_MANAGEMENT_TASKS.required_decorator()], name='dispatch')
+class DataManagementView(BaseProjectDataView):
+    urlname = 'data_management'
+    page_title = "Data Management"
+    template_name = 'data_management/data_management.html'
+
+    @use_jquery_ui  # for datepicker
```
Comment (snopoke) [inline custom/icds/data_management/views.py]: why not add this to the list of decorators above?

File: custom/icds/data_management/doc_processors.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        for case_id, new_value in updates.items():
+            case_block = CaseBlock(case_id,
+                                   update={case_property: new_value},
+                                   user_id=SYSTEM_USER_ID)
+            case_block = ElementTree.tostring(case_block.as_xml()).decode('utf-8')
+            case_blocks.append(case_block)
+        return case_blocks
+
+    def handle_skip(self, doc):
+        print('Unable to process case {}'.format(doc['_id']))
```
Comment (snopoke) [inline custom/icds/data_management/doc_processors.py]: who will see this message?

File: custom/icds/data_management/base.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+class DataManagement(object):
+    slug = ""
+    name = ""
+    case_type = None
+    doc_processor = None
+
+    def __init__(self, domain, *args, **kwargs):
+        self.domain = domain
+
+    def case_accessor(self):
```
Comment (snopoke) [inline custom/icds/data_management/base.py]: why is this a function and not a class property like the other fields?

File: custom/icds/data_management/base.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        raise NotImplementedError
+
+    def run(self, iteration_key):
+        """
+        iterate sql records and update them as and when needed
+        """
+        record_provider = SqlDocumentProvider(iteration_key, self.case_accessor())
+        processor = BulkDocProcessor(record_provider, self.doc_processor(self.domain))
+        processor.run()
```
Comment (snopoke) [inline custom/icds/data_management/base.py]: I would recommend recording the values returned from here `num_processes, num_skipped`

**Review by snopoke — CHANGES_REQUESTED** [review]
Main concerns:
* what happens to documents that fail, it doesn't appear that the operator would be able to even see how many were processed vs skipped
* some of these operations are very expensive, is there a parallel process to fix the root causes?

**Comment by mkangia** [comment]
@snopoke 
> what happens to documents that fail, it doesn't appear that the operator would be able to even see how many were processed vs skipped

I would like to do that as a follow up where we notify the missed documents or any sort of logging stored and shared with the user for cases that were missed/done. There is actually more to it where we want to store a certain type of cases as well like in case of phone number updates we also want to note down where user has an unexpected phone number. So since its a big chunk, i will take it in a follow up PR

> some of these operations are very expensive, is there a parallel process to fix the root causes?
Not sure what you are referring to here, we do club updates to cases. Is there any other place were you see there can be an improvement.
In terms of number of cases, Since these would be done every monthly or quarterly, the number of cases iterated would be less. We can add a restriction to "from date" or "start date" (post change suggested) to be within a certain limit.

File: custom/icds/data_management/doc_processors.py (hunk 6 above)
Comment (mkangia) [reply custom/icds/data_management/doc_processors.py]: Currently just logged for testing, would be enhanced to share these case ids with the user initiating the request when we take up feedback to user in a follow up PR.

File: custom/icds/data_management/views.py (hunk 5 above)
Comment (mkangia) [reply custom/icds/data_management/views.py]: `use_jquery_ui` does not support that. I think i tried modifying it to be applied over class earlier as well but it failed.

File: custom/icds/data_management/models.py (hunk 3 above)
Comment (mkangia) [reply custom/icds/data_management/models.py]: We are updating the record itself, so to refresh it? Even though it might not appear to be useful, but just wanted to get any changes that might happen on save like `created_at` would get set on this save.

File: corehq/form_processor/backends/sql/dbaccessors.py (hunk 1 above)
Comment (mkangia) [reply corehq/form_processor/backends/sql/dbaccessors.py]: I was not sure how to integrate that with rest of the structure. My understanding was that classes inherited from `ReindexAccessor` have a certain structure that is needed to use sql iterator?

File: custom/icds/data_management/base.py (hunk 7 above)
Comment (mkangia) [reply custom/icds/data_management/base.py]: You mean something like this?
```
class DataManagement(object):
    case_accessor = None

class SQLBasedDataManagement(DataManagement):
    case_accessor = CaseReindexAccessor

    def run(self, iteration_key):
        record_provider = SqlDocumentProvider(iteration_key, self.case_accessor(
            domain=self.domain,
            case_type=self.case_type,
            limit_db_aliases=self.db_aliases,
            start_date=self.from_date,
            end_date=self.till_date
        ))
```

Considering the arguments passed, It's going to change for each class and each instance of class. Keeping it on instance made `run` method generic and i could keep that at one place. 
I can still keep it as a property on instance, but considered not caching it because its not reused, and its dependent on the properties set on the instant which can change if someone is running these requests manually

File: custom/icds/data_management/models.py (hunk 4 above)
Comment (mkangia) [reply custom/icds/data_management/models.py]: `execute` is the only public method/entry point for this class, the form already ensures that no other value is entered in the UI but kept the check on the model as well.
Where before this you suggest it should be checked?

File: custom/icds/data_management/views.py (hunk 5 above)
Comment (snopoke) [reply custom/icds/data_management/views.py]: that's a shame. I see it has specifically been built as a method decorator

File: custom/icds/data_management/base.py (hunk 2 above)
Comment (mkangia) [reply custom/icds/data_management/base.py]: https://github.com/dimagi/commcare-hq/pull/26610/commits/d4214e80d4f871bd032ac00fe93345f35bd092ac

File: custom/icds/data_management/base.py (hunk 8 above)
Comment (mkangia) [reply custom/icds/data_management/base.py]: https://github.com/dimagi/commcare-hq/pull/26610/commits/c1630e9fb4ec26d0171c0cc8e7b815e83bb7a961

It's not being used for now but still returned. We can use it in the follow up PR when making feedback for the user

File: custom/icds/data_management/models.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        self.save()
+        self.refresh_from_db()
+
+    def _note_error(self, message):
+        self.error = message
+        self.status = self.STATUS_FAILED
+
+    def _perform_task(self):
+        task_to_run = DATA_MANAGEMENT_TASKS[self.slug](self.domain, self.db_alias, self.start_date, self.end_date)
+        iteration_key = "%s-%s-%s-%s-%s" % (self.slug, self.start_date, self.end_date, self.initiated_by,
```
Comment (snopoke) [inline custom/icds/data_management/models.py]: why is `end_date` in here? won't it always be `None` at this point?

File: custom/icds/data_management/models.py (hunk 4 above)
Comment (snopoke) [reply custom/icds/data_management/models.py]: fair enough

File: custom/icds/data_management/models.py (hunk 3 above)
Comment (snopoke) [reply custom/icds/data_management/models.py]: I'm almost 100% sure that any values updated in the DB will be set on the model during the save call. I think this is only needed if you expect there to be changes directly in the DB or else from a separate instance of the model.

File: custom/icds/data_management/models.py (hunk 9 above)
Comment (mkangia) [reply custom/icds/data_management/models.py]: hmm, not always. User can choose to work on cases modified between a duration in which case both start and end date would be set.

File: custom/icds/data_management/models.py (hunk 3 above)
Comment (mkangia) [reply custom/icds/data_management/models.py]: okay cool. Just was thinking of a request instance being rerun and modified in parallel in another process, an edge case not worth refetching record for.
https://github.com/dimagi/commcare-hq/pull/26610/commits/0802b61dc7c6c1398e02f9baef600f338bc4a7ae

File: corehq/form_processor/backends/sql/dbaccessors.py (hunk 1 above)
Comment (snopoke) [reply corehq/form_processor/backends/sql/dbaccessors.py]: You could just create a new [DocumentProvider](https://github.com/dimagi/commcare-hq/blob/4a6f4cede4c01f482fe9dc9857e754c65f433325/corehq/util/doc_processor/interface.py#L61) class that takes a list of case_ids.

Comment (mkangia) [reply corehq/form_processor/backends/sql/dbaccessors.py]: https://github.com/dimagi/commcare-hq/pull/26610/commits/90d36e8b9e4004bf89f277a654db60eb2962cd4a

**Review by snopoke — APPROVED** [review]
